### PR TITLE
Implement RegularLanguageLexer

### LINKED ISSUE #38: Implement Lexer

### PR DESCRIPTION
**Pull request by km1chno**
Resolves #38 

## Changes

- change `Lexer` and `Diagnostics` to interface
- implement `RegularLanguageLexer` implementing `Lexer` interface + unit tests
- add some files with constants

## TODO

- ~wait for implementation of Regex->DFA to implement `RegularLanguageLexer::fromRegexes()`~
- ~wait for implementation of StringInput and FileInput to get rid of `createStringInput()` from `TestUtils`~
- ~think of message to diagnostics on lexing fail~
- ~what is `Token::context` supposed to be?~

### CODE REVIEW

File: src/main/kotlin/cacophony/Constants.kt [hunk 1]
```diff
@@ -0,0 +1,4 @@
+package cacophony
```
Comment (Giselus) [inline src/main/kotlin/cacophony/Constants.kt:1]: Good idea to actually create this file, we may later fix other places to get comment_start from there.

File: src/test/kotlin/cacophony/TestUtils.kt [hunk 2]
```diff
@@ -0,0 +1,69 @@
+package cacophony
+
+import cacophony.automata.DFA
+import cacophony.utils.Input
+import cacophony.utils.Location
+
+fun createDFA(
```
Comment (Giselus) [inline src/test/kotlin/cacophony/TestUtils.kt:7]: There is simpleDFA that does exactly this. Maybe it's still better to mock it, but seems weird if code is almost the same.

File: src/test/kotlin/cacophony/TestUtils.kt [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        override fun getProduction(
+            state: Int,
+            symbol: Char,
+        ): Int? = productions[Pair(state, symbol)]
+
+        override fun isAccepting(state: Int): Boolean = accepting.contains(state)
+    }
+}
+
+// TODO: This can be removed once Rafał implements inputs
```
Comment (Giselus) [inline src/test/kotlin/cacophony/TestUtils.kt:35]: Resolve TODO ofc.

File: src/test/kotlin/cacophony/TestConstants.kt
Comment (Giselus) [inline src/test/kotlin/cacophony/TestConstants.kt:1]: I'm not sure if you should create consts like this if you want to use them in only one test file.

**Review by Giselus — CHANGES_REQUESTED** [review]
Looks good to me, please fix all of the places regarding token context.

File: src/main/kotlin/cacophony/lexer/RegularLanguageLexer.kt [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+                        }
+                    } ?: run {
+                        input.next()
+                        diagnostics.report("No match found", input, input.getLocation()) // TODO: Better error message
+                    }
+                }
+            }
+        }
+
+        return tokens
```
[the remark this answers is not in the document]
Comment (km1chno) [reply src/main/kotlin/cacophony/lexer/RegularLanguageLexer.kt:108]: im not sure I understand, do you mean the variable "tokens" is an empty list or that we find empty tokens? Can we find empty tokens?

File: src/test/kotlin/cacophony/TestConstants.kt
Comment (km1chno) [reply src/test/kotlin/cacophony/TestConstants.kt:1]: I thought they might be used in other test suites in the future, wdyt? if its not likely, I can move them to lexer tests file

File: src/main/kotlin/cacophony/lexer/RegularLanguageLexer.kt [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+        while (true) {
+            when (input.peek()) {
+                COMMENT_START_CHAR -> input.skip(END_OF_LINE_CHAR)
+                null -> break
+                else -> {
+                    matchNewToken(input)?.let { token ->
+                        run {
+                            tokens.add(token)
+                            input.setLocation(token.rangeTo)
```
[the remark this answers is not in the document]
Comment (km1chno) [reply src/main/kotlin/cacophony/lexer/RegularLanguageLexer.kt]: true, thanks

File: src/main/kotlin/cacophony/lexer/RegularLanguageLexer.kt [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+                        }
+                    } ?: run {
+                        input.next()
+                        diagnostics.report("No match found", input, input.getLocation()) // TODO: Better error message
```
[the remark this answers is not in the document]
Comment (km1chno) [reply src/main/kotlin/cacophony/lexer/RegularLanguageLexer.kt]: fixed

File: src/test/kotlin/cacophony/TestUtils.kt (hunk 2 above)
Comment (km1chno) [reply src/test/kotlin/cacophony/TestUtils.kt]: nice, I changed it to use simpleDFA. Also went ahead and did the same in `DFAEquivalenceTest.kt`

File: src/test/kotlin/cacophony/TestUtils.kt (hunk 3 above)
Comment (km1chno) [reply src/test/kotlin/cacophony/TestUtils.kt]: done

File: src/main/kotlin/cacophony/lexer/RegularLanguageLexer.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (Giselus) [reply src/main/kotlin/cacophony/lexer/RegularLanguageLexer.kt:98]: I just had stroke writing this I think. I guess I meant that maybe we want to return empty list if we got some lexing error along the way, but I'm not sure, maybe now it's okay.

[the remark this answers is not in the document]
Comment (km1chno) [reply src/main/kotlin/cacophony/lexer/RegularLanguageLexer.kt:105]: nah I'd leave it as is

File: src/main/kotlin/cacophony/Constants.kt (hunk 1 above)
Comment (km1chno) [reply src/main/kotlin/cacophony/Constants.kt:1]: imo not in this pr though, looking at our code i think some refactoring is inevitable...

**Review by Giselus — APPROVED** [review]

**Pull request merged**